pkg/executor/join/anti_semi_join_probe.go

```diff
@@ -0,0 +1,337 @@
+		if err != nil {
+			return err
+		}
+
+		for index, result := range a.selected {
+			if result || a.isNulls[index] {
+				candidateRow := unsafe.Pointer(&a.rowIndexInfos[index].buildRowStart)
+				if !meta.isCurrentRowUsedWithAtomic(candidateRow) {
+					meta.setUsedFlag(*(*unsafe.Pointer)(candidateRow))
+				}
+			}
+		}
+	}
+
+	return
+}
+
+func (a *antiSemiJoinProbe) probeForLeftSideBuildNoOtherCondition(sqlKiller *sqlkiller.SQLKiller) (err error) {
+	meta := a.ctx.hashTableMeta
+	tagHelper := a.ctx.hashTableContext.tagHelper
+
+	loopCnt := 0
+
+	for a.currentProbeRow < a.chunkRows {
+		if a.matchedRowsHeaders[a.currentProbeRow] != 0 {
+			candidateRow := tagHelper.toUnsafePointer(a.matchedRowsHeaders[a.currentProbeRow])
+			if !meta.isCurrentRowUsedWithAtomic(candidateRow) {
+				if isKeyMatched(meta.keyMode, a.serializedKeys[a.currentProbeRow], candidateRow, meta) {
+					meta.setUsedFlag(candidateRow)
+				} else {
+					a.probeCollision++
+				}
+			}
+			a.matchedRowsHeaders[a.currentProbeRow] = getNextRowAddress(candidateRow, tagHelper, a.matchedRowsHashValue[a.currentProbeRow])
+		} else {
+			a.currentProbeRow++
+		}
+
+		loopCnt++
+		if loopCnt%2000 == 0 {
+			err = checkSQLKiller(sqlKiller, "killedDuringProbe")
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = checkSQLKiller(sqlKiller, "killedDuringProbe")
+	return err
+}
+
+func (a *antiSemiJoinProbe) produceResult(joinedChk *chunk.Chunk, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	err = a.concatenateProbeAndBuildRows(joinedChk, sqlKiller, true)
+	if err != nil {
+		return err
+	}
+
+	if joinedChk.NumRows() > 0 {
+		a.selected, a.isNulls, err = expression.VecEvalBool(a.ctx.SessCtx.GetExprCtx().GetEvalCtx(), a.ctx.SessCtx.GetSessionVars().EnableVectorizedExpression, a.ctx.OtherCondition, joinedChk, a.selected, a.isNulls)
+		if err != nil {
+			return err
+		}
+
+		length := len(a.selected)
+		for i := range length {
+			if a.selected[i] {
+				probeRowIdx := a.rowIndexInfos[i].probeRowIndex
+				a.isMatchedRows[probeRowIdx] = true
+			}
+
+			if a.isNulls[i] {
+				probeRowIdx := a.rowIndexInfos[i].probeRowIndex
+				a.isMatchedRows[probeRowIdx] = true
+			}
+		}
+	}
+	return
+}
+
+func (a *antiSemiJoinProbe) probeForRightSideBuildHasOtherCondition(chk, joinedChk *chunk.Chunk, remainCap int, sqlKiller *sqlkiller.SQLKiller) (err error) {
+	if !a.unFinishedProbeRowIdxQueue.IsEmpty() {
+		err = a.produceResult(joinedChk, sqlKiller)
+		if err != nil {
+			return err
+		}
+		a.currentProbeRow = 0
+	}
+
+	if a.unFinishedProbeRowIdxQueue.IsEmpty() {
+		if !a.outputProbeRowsSelected {
+			// As this is anti semi, only rows that marked as not matched are result rows.
+			for i, matched := range a.isMatchedRows {
+				a.isMatchedRows[i] = !matched
+			}
+
+			a.outputProbeRowsSelected = true
```

Comment:
maybe rename it as `isMatchedRowsAdjustedForOutput`

Reply:
> maybe rename it as `isMatchedRowsAdjustedForOutput`

done